light-clients/icsxx-cf-solana/src/client_def.rs

```diff
@@ -48,7 +49,51 @@ where
 		client_state: Self::ClientState,
 		client_msg: Self::ClientMessage,
 	) -> Result<(), Ics02ClientError> {
-		client_state.0.do_verify_client_message(self, client_msg.0).map_err(convert)
+		match client_msg {
+			ClientMessage::Header(header) => {
+				// The client can't be updated if no shreds were received
+				let shreds = header.shreds;
+				if shreds.is_empty() {
+					return Err(Ics02ClientError::implementation_specific(
+						"no shreds received".to_string(),
+					));
+				}
+
+				let slot = shreds.first().unwrap().slot();
+
+				// All shreds' slots should be the same
+				shreds.iter().skip(1).all(|s| s.slot() == slot).then(|| ()).ok_or_else(|| {
+					Ics02ClientError::implementation_specific(
+						"shreds have different slots".to_string(),
+					)
+				})?;
+
+				if slot <= client_state.latest_height().revision_height {
+					return Err(Ics02ClientError::implementation_specific(
+						"slot is not greater than latest height".to_string(),
+					));
+				}
+
+				let leader = client_state.leader_for_slot(slot);
+
+				// Verify all shreds
+				shreds
+					.iter()
+					.map(|shred| shred.sanitize().and_then(|_| shred.verify_with_root(&leader)))
+					.collect::<Result<Vec<_>, _>>()
+					.map_err(|e| {
+						Ics02ClientError::implementation_specific(alloc::format!(
+							"shred verification failed: {}",
+							e
+						))
+					})?;
+				Ok(())
```

Comment:
Since you’re not doing anything with the shards collecting them into
Vec is waste of memory.

```suggestion
				header.shreds.iter().try_for_each(|shred| {
					shred.sanitize()?;
					shred.verify_with_root(&leader)
				}).map_err(|err| {
					Ics02ClientError::implementation_specific(alloc::format!(
						"shred verification failed: {err}",
					))
				})
```

Reply:
Good call!